File: src/services/githubService.js
const axios = require("axios");
const NodeCache = require("node-cache");

const cache = new NodeCache({ stdTTL: 300 }); // Cache for 5 minutes

const githubAPI = axios.create({
  baseURL: "https://api.github.com",
  headers: {
    Accept: "application/vnd.github+json",
    ...(process.env.GITHUB_TOKEN && {
      Authorization: `Bearer ${process.env.GITHUB_TOKEN}`,
    }),
  },
});

async function cachedGet(key, fetchFn) {
  const cached = cache.get(key);
  if (cached) return cached;
  const data = await fetchFn();
  cache.set(key, data);
  return data;
}

exports.fetchProfile = async (username) => {
  return cachedGet(`profile:${username}`, async () => {
    const { data } = await githubAPI.get(`/users/${username}`);
    return {
      login: data.login,
      name: data.name,
      avatar: data.avatar_url,
      bio: data.bio,
      location: data.location,
      blog: data.blog,
      followers: data.followers,
      following: data.following,
      publicRepos: data.public_repos,
      createdAt: data.created_at,
      url: data.html_url,
    };
  });
};

exports.fetchRepos = async (username, sort = "updated", limit = 6) => {
  return cachedGet(`repos:${username}:${sort}:${limit}`, async () => {
    const { data } = await githubAPI.get(`/users/${username}/repos`, {
      params: { sort, per_page: limit },
    });
    return data.map((repo) => ({
      name: repo.name,
      description: repo.description,
      stars: repo.stargazers_count,
      forks: repo.forks_count,
      language: repo.language,
      url: repo.html_url,
      updatedAt: repo.updated_at,
      topics: repo.topics || [],
    }));
  });
};

exports.fetchActivity = async (username) => {
  return cachedGet(`activity:${username}`, async () => {
    const { data } = await githubAPI.get(`/users/${username}/events/public`, {
      params: { per_page: 10 },
    });
    return data.map((event) => ({
      type: event.type,
      repo: event.repo.name,
      createdAt: event.created_at,
      payload: summarizePayload(event),
    }));
  });
};

exports.fetchStats = async (username) => {
  return cachedGet(`stats:${username}`, async () => {
    const { data: repos } = await githubAPI.get(`/users/${username}/repos`, {
      params: { per_page: 100 },
    });

    const languages = {};
    let totalStars = 0;
    let totalForks = 0;

    repos.forEach((repo) => {
      totalStars += repo.stargazers_count;
      totalForks += repo.forks_count;
      if (repo.language) {
        languages[repo.language] = (languages[repo.language] || 0) + 1;
      }
    });

    const topLanguages = Object.entries(languages)
      .sort((a, b) => b[1] - a[1])
      .slice(0, 5)
      .map(([lang, count]) => ({ lang, count }));

    return { totalStars, totalForks, topLanguages, totalRepos: repos.length };
  });
};

function summarizePayload(event) {
  switch (event.type) {
    case "PushEvent":
      return `Pushed ${event.payload.commits?.length || 0} commit(s)`;
    case "CreateEvent":
      return `Created ${event.payload.ref_type} ${event.payload.ref || ""}`;
    case "WatchEvent":
      return "Starred a repository";
    case "ForkEvent":
      return "Forked a repository";
    case "IssuesEvent":
      return `${event.payload.action} an issue`;
    case "PullRequestEvent":
      return `${event.payload.action} a pull request`;
    default:
      return event.type.replace("Event", "");
  }
}

File: src/services/github_service.py
"""
GitHub Service
Handles all GitHub API requests with in-memory caching (5-minute TTL).
Equivalent to src/services/githubService.js
"""

import os
import time
import requests
from threading import Lock

GITHUB_API = "https://api.github.com"
CACHE_TTL = 300  # 5 minutes

_cache: dict = {}
_lock = Lock()


def _get_headers() -> dict:
    token = os.getenv("GITHUB_TOKEN")
    headers = {"Accept": "application/vnd.github+json"}
    if token:
        headers["Authorization"] = f"Bearer {token}"
    return headers


def _cache_get(key: str):
    with _lock:
        entry = _cache.get(key)
        if entry and (time.time() - entry["ts"]) < CACHE_TTL:
            return entry["data"]
        return None


def _cache_set(key: str, data):
    with _lock:
        _cache[key] = {"data": data, "ts": time.time()}


def _github_get(path: str) -> dict | list:
    """Make a GET request to the GitHub API, raising on error."""
    url = f"{GITHUB_API}{path}"
    response = requests.get(url, headers=_get_headers(), timeout=10)
    if response.status_code == 404:
        raise ValueError(f"GitHub user or resource not found: {path}")
    if response.status_code == 403:
        raise PermissionError("GitHub API rate limit exceeded. Add a GITHUB_TOKEN to your .env file.")
    response.raise_for_status()
    return response.json()


# ── Public API ────────────────────────────────────────────────────────────────

def get_profile(username: str) -> dict:
    key = f"profile:{username}"
    cached = _cache_get(key)
    if cached:
        return cached

    data = _github_get(f"/users/{username}")
    result = {
        "login": data.get("login"),
        "name": data.get("name"),
        "avatar_url": data.get("avatar_url"),
        "bio": data.get("bio"),
        "location": data.get("location"),
        "company": data.get("company"),
        "blog": data.get("blog"),
        "public_repos": data.get("public_repos"),
        "followers": data.get("followers"),
        "following": data.get("following"),
        "created_at": data.get("created_at"),
        "html_url": data.get("html_url"),
    }
    _cache_set(key, result)
    return result


def get_repos(username: str, sort: str = "updated", limit: int = 6) -> list:
    key = f"repos:{username}:{sort}:{limit}"
    cached = _cache_get(key)
    if cached:
        return cached

    data = _github_get(f"/users/{username}/repos?sort={sort}&per_page=100&type=owner")
    repos = []
    for r in data[:limit]:
        repos.append({
            "id": r.get("id"),
            "name": r.get("name"),
            "full_name": r.get("full_name"),
            "description": r.get("description"),
            "html_url": r.get("html_url"),
            "language": r.get("language"),
            "stargazers_count": r.get("stargazers_count"),
            "forks_count": r.get("forks_count"),
            "topics": r.get("topics", []),
            "updated_at": r.get("updated_at"),
            "created_at": r.get("created_at"),
        })
    _cache_set(key, repos)
    return repos


def get_activity(username: str) -> list:
    key = f"activity:{username}"
    cached = _cache_get(key)
    if cached:
        return cached

    data = _github_get(f"/users/{username}/events/public?per_page=30")
    events = []
    for e in data:
        events.append({
            "id": e.get("id"),
            "type": e.get("type"),
            "repo": e.get("repo", {}).get("name"),
            "created_at": e.get("created_at"),
            "payload_summary": _summarise_payload(e),
        })
    _cache_set(key, events)
    return events


def get_stats(username: str) -> dict:
    key = f"stats:{username}"
    cached = _cache_get(key)
    if cached:
        return cached

    repos_raw = _github_get(f"/users/{username}/repos?per_page=100&type=owner")

    lang_counts: dict[str, int] = {}
    total_stars = 0
    total_forks = 0

    for r in repos_raw:
        lang = r.get("language")
        if lang:
            lang_counts[lang] = lang_counts.get(lang, 0) + 1
        total_stars += r.get("stargazers_count", 0)
        total_forks += r.get("forks_count", 0)

    sorted_langs = sorted(lang_counts.items(), key=lambda x: x[1], reverse=True)

    result = {
        "total_repos": len(repos_raw),
        "total_stars": total_stars,
        "total_forks": total_forks,
        "languages": [{"name": l, "count": c} for l, c in sorted_langs],
    }
    _cache_set(key, result)
    return result


# ── Helpers ───────────────────────────────────────────────────────────────────

def _summarise_payload(event: dict) -> str:
    t = event.get("type", "")
    p = event.get("payload", {})
    repo = event.get("repo", {}).get("name", "")

    if t == "PushEvent":
        commits = len(p.get("commits", []))
        return f"Pushed {commits} commit(s) to {repo}"
    if t == "PullRequestEvent":
        action = p.get("action", "")
        pr = p.get("pull_request", {}).get("title", "")
        return f"{action.capitalize()} PR: {pr}"
    if t == "IssuesEvent":
        action = p.get("action", "")
        title = p.get("issue", {}).get("title", "")
        return f"{action.capitalize()} issue: {title}"
    if t == "WatchEvent":
        return f"Starred {repo}"
    if t == "ForkEvent":
        return f"Forked {repo}"
    if t == "CreateEvent":
        ref_type = p.get("ref_type", "")
        ref = p.get("ref", "")
        return f"Created {ref_type} {ref} in {repo}"
    if t == "DeleteEvent":
        ref_type = p.get("ref_type", "")
        return f"Deleted {ref_type} in {repo}"
    return t.replace("Event", "")
